### PR TITLE
DOC: update Ubuntu quickstart; conda compilers now work!

### PR DESCRIPTION
**Pull request by mdhaber**
The wait is over; conda compilers now seem to build SciPy.

Tested on fresh instances of Ubuntu 16.04 and 18.04  EC2 instance and 20.04 WSL locally. That's all I can do, but please test on your system if possible.

### CODE REVIEW

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst:51]: @rgommers Does `mypy` help? Is there a reason to install `sphinx` without `matplotlib`?

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (ev-br) [inline doc/source/dev/contributor/quickstart_ubuntu.rst:76]: Why an inplace build here?  I think `python setup.py build && python runtests.py` builds scipy only once. 

Anyway, it's a small matter, feel free to ignore

Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:76]: If you don't do an in-place build, don't you have to build every time you want to test changes?

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:51]: This renders as:
image

Is the inconsistent styling avoidable?

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (ev-br) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:76]: Not with `runtests.py`, I think. Subsequent rebuilds are much faster, so I suspect the sys.path manipulations in runtests are equivalent to an inplace build.

Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:76]: What if you are running a script other than `runtests.py` or testing changes interactively in an IDE?

**Comment by sethtroisi** [comment]
Thanks for updating the docs. always nice when they are fresh and up to date :+1:

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst:6]: Can someone try this as-is on Mac? Now that `conda` compilers seem to be working, there may not be a need for the Homebrew compilers, and maybe we can unify the two guides.

Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:6]: According to [this](https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html), the compilers for Mac are `clang_osx-64`, `clangxx_osx-64`, and `gfortran_osx-64`. I modified the `conda install` command accordingly, but build failed.
```
ld: could not parse object file build/temp.macosx-10.9-x86_64-3.8/scipy/integrate/_quadpackmodule.o: 'Unknown attribute kind (61) (Producer: 'LLVM10.0.0' Reader: 'LLVM APPLE_1_1000.10.44.4_0')', using libLTO version 'LLVM version 10.0.0, (clang-1000.10.44.4)' file 'build/temp.macosx-10.9-x86_64-3.8/scipy/integrate/_quadpackmodule.o' for architecture x86_64
collect2: error: ld returned 1 exit status
error: Command "/Users/matthaberland/anaconda3/envs/scipydev/bin/x86_64-apple-darwin13.4.0-gfortran -Wall -g -Wall -g -undefined dynamic_lookup -bundle -Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,/Users/matthaberland/anaconda3/envs/scipydev/lib -L/Users/matthaberland/anaconda3/envs/scipydev/lib build/temp.macosx-10.9-x86_64-3.8/scipy/integrate/_quadpackmodule.o -L/Users/matthaberland/anaconda3/envs/scipydev/lib -L/Users/matthaberland/anaconda3/envs/scipydev/bin/../lib/gcc/x86_64-apple-darwin11.4.2/4.8.5 -L/Users/matthaberland/anaconda3/envs/scipydev/bin/../lib/gcc/x86_64-apple-darwin11.4.2/4.8.5/../../.. -L/Users/matthaberland/anaconda3/envs/scipydev/bin/../lib/gcc/x86_64-apple-darwin11.4.2/4.8.5/../../.. -Lbuild/temp.macosx-10.9-x86_64-3.8 -lquadpack -lmach -lmkl_rt -lpthread -lgfortran -o scipy/integrate/_quadpack.cpython-38-darwin.so" failed with exit status 1
```

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (rgommers) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:51]: > Does `mypy` help? Is there a reason to install `sphinx` without `matplotlib`?

The aim is to get a complete development environment here. So yes I'd like to add `mypy`. And good point about `matplotlib`, let's add that for doc building.

> Yes, this can be combined with creating the environment, and that is what an experienced user should do. But someone new to `conda` would benefit from understanding what these commands do separately.

Actually it is better practice to install everything at once. You're not guaranteed to get the same optimal solution if you chop it up.

Comment (rgommers) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:51]: Re doc styling: you need to unindent the bullet list. The vertical bar on the left indicates that you added unwanted indentation, and Sphinx tries to render as a quote.

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (rgommers) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:76]: > or testing changes interactively in an IDE? 

I think that's what `runtests.py --ipython` is for. In an IDE you'd have to set PYTHONPATH manually.

I think `runtests.py` has pros and cons; I've never been much of a fan, but it does make life easier for beginning contributors for most development tasks.

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (rgommers) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:6]: I think on macOS combining those compilers doesn't work without `conda-build`.

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (peterbell10) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:76]: I thought the point of `--inplace` is you can import extension modules directly in the project directory without installing the whole library. However, SciPy expressly forbids doing that:

https://github.com/scipy/scipy/blob/28d4244b6f5c1543755e586f27f9261d0f84153b/scipy/__init__.py#L122-L124

If that's right, then I don't think you'd ever want to build SciPy inplace.

If you're working on native code `runtests.py` is excellent because you can't forget to rebuild before testing. If you only work on python then I guess `python setup.py develop` would be just as good?

Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:76]: OK. I can't make that change, though, because I don't know about this feature of `runtests.py`.

Comment (ev-br) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:78]: TBH `runtests.py --ipython` is so huge relief compared to fiddling with PYTHONPATH or `setup.py develop` or forgetting to `setup.py install` etc

workflow wise I typically do inplace builds for small projects. When dealing with beginners, `setup.py build --i place` is generally a great suggestion --- with an explicit exception of building numpy/scipy where runtests.py is already set up.

So I'd suggest recommending runtests in the beginner quickstart guide for building scipy specifically, and mention inplace in a more general "working with compiled extensions" part of the guide.

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:51]: > Re doc styling: you need to unindent the bullet list

I didn't see the extra space; it needed three instead of four.

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:78]: > In an IDE you'd have to set PYTHONPATH manually.

`conda develop .`

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:51]: > So yes I'd like to add `mypy`.

OK. I've never used it, so I don't know what to write about it.

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (rgommers) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:78]: > I thought the point of `--inplace` is you can import extension modules directly in the project directory without installing the whole library. However, SciPy expressly forbids doing that:

No that's not it. My typical workflow is `python setup.py build_ext -i` followed by (in another shell from anywhere, or from an editor/IDE, with `PYTHONPATH` pointing to repo root) importing and testing. It's faster iterating than with `runtests.py`, and one can use any tool one wants rather than being limited to what `runtests.py` supports as a command-line flag.

Comment (peterbell10) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:78]: @rgommers what benefit do you get specifically from the `-i` flag though?

Comment (rgommers) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:78]: 1. That I know where my build is (typically `~/code/scipy`), rather than in some new directory that `runtests.py` created
2. If I want to edit Python code in SciPy, I can just have edit-test cycles (e.g. in a Jupyter notebook), rather than "edit, run-runtests-to-copy-edited-code-over, test" cycles.
3. I want to see my build log as it goes

To me, `runtests.py` is nice for beginners because it abstracts some details. Once you know those details though, the abstraction layer just gets in the way.

Comment (peterbell10) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:78]: 1: It's not `runtests.py` that creates the folder, it's `distutils`. Out of source builds are the default because it's easier to clean up build artefacts and you can build multiple incompatible configurations without interfering with the other's build artefacts. e.g. building for multiple python versions in the same directory without cleaning in between.

2 & 3: The same is true for `setup.py develop`, no?

Comment (rgommers) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:78]: 1. Well, that's semantics - `runtests.py` has extensive `PYTHONPATH` manipulation and building with `'--prefix=' , '--single-version-externally-managed',` to control where the build ends up.

> 2 & 3: The same is true for setup.py develop, no?

Yes exactly. `build_ext -i` and `develop` are effectively the same, the only difference is that `develop` makes an entry in `site-packages` so you should use that if you have a separate development environment only, while `build_ext -i` you can use even in your main env where you have Jupyter, Dask and whatever else you want to play with installed. I tend to do the latter more than the former, since remembering to do `conda activate scipy-dev` each time I open a new terminal is a pain.

Comment (peterbell10) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:78]: Ah so maybe recommend `develop` to people using environments and `build_ext -i` if using the system python?

Comment (ev-br) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:78]: As long as it's a quickstart tutorial, I guess it's best to pick _a_ recommendation and stick to it. And link to a separate "possible workflows" document which lists various workflows.

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:6]: After installing `conda-build`, the problem persisted.

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:78]: The intent of this PR was to make simplifications based on the conda compilers working. This PR didn't add the `--inplace` option, so unless there is consensus, maybe we leave this to a separate issue/PR?

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (peterbell10) [inline doc/source/dev/contributor/quickstart_ubuntu.rst:33]: Should probably remove Spyder from here if you're removing it from later on as well.

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (peterbell10) [inline doc/source/dev/contributor/quickstart_ubuntu.rst:35]: Not strictly related to this PR, but IMO you should either give the user a complete curl command to download `conda` just in terminal, or give them a link to the download page on the anaconda website. Giving them a link to follow in their browser to find the correct URL to pass to curl just seems confusing.

Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:35]: I think the filename changes as new versions are released, so I don't think we can provide the complete command. 

Yes, in many cases, the `curl` command would be unnecessary. In many cases, if they can browse to the website, they can also download in their browser and use a graphical installer. I included the `curl` command for WSL and remote terminal users who can browse to the site using their native OS but can only interact via terminal with the machine on which they're doing the installation. That's what I do.

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst:33]:
```suggestion
#. Download, install, and test the latest release of the `Anaconda Distribution of Python`_. In addition to the latest version of Python 3, the Anaconda distribution includes dozens of the most popular Python packages for scientific computing, the ``conda`` package manager, and tools for managing virtual environments.
```

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (peterbell10) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:35]: I would be in favour of leaving out curl then. I think it's reasonable to assume that anyone using a remote terminal knows how to transfer a file from their local machine.

Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:35]: I didn't, and that's why I included it. But I'll take it out.

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst:35]:
```suggestion
```

Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:35]: Should we keep "If you're installing using the terminal, be sure to follow the "Next Steps" listed after the installer finishes. You might also need to restart your terminal window or enter ``source ~/.bashrc`` for all the changes to take effect."?

Comment (peterbell10) [reply doc/source/dev/contributor/quickstart_ubuntu.rst:35]: Yeah, that part's all good.

Comment (mdhaber) [reply doc/source/dev/contributor/quickstart_ubuntu.rst]: OK, all set.

**Review by peterbell10 — APPROVED** [review]

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst:54]:
```suggestion
#. Enter ``conda install python=3.8 numpy pybind11 cython pytest gfortran_linux-64 gxx_linux-64 sphinx matplotlib mypy git``. |br| Note that we're installing SciPy's build dependencies and some other software, but not SciPy itself.
```

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst:62]:
```suggestion
   .. note::
```

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst:68]:
```suggestion
      You could ``conda create`` an empty virtual environment first, then
      ``conda install`` the packages, but creating the virtual environment
      with all the packages you need is preferable to installing packages
      individually because it makes it easier for ``conda`` to solve
      the package dependencies optimally.
```

File: doc/source/dev/contributor/quickstart_ubuntu.rst
Comment (mdhaber) [inline doc/source/dev/contributor/quickstart_ubuntu.rst:70]:
```suggestion
#. You're still in the base environment. Activate your new virtual environment by entering ``conda activate scipydev``. |br| If you're working with an old version of ``conda``, you might need to type ``source activate scipydev`` instead (see `here <https://stackoverflow.com/questions/49600611/python-anaconda-should-i-use-conda-activate-or-source-activate-in-linux)>`__). Note that you'll need to have this virtual environment active whenever you want to work with the development version of SciPy.
```

**Comment by mdhaber** [comment]
@rgommers I saw you mentioned the Ubuntu instructions in gh-12900; they will be even simpler once this is merged. I think I addressed all your comments.

**Review by rgommers — APPROVED** [review]
This is an improvement, so let's get it in. Thanks @mdhaber for the update, and everyone for the reviews.

**Pull request merged**
